internal/executor/piper/piper.go

```diff
@@ -0,0 +1,67 @@
+package piper
+
+import (
+	"errors"
+	"io"
+	"os"
+	"runtime"
+	"time"
+)
+
+type Piper struct {
+	r, w    *os.File
+	errChan chan error
+}
+
+func New(output io.Writer) (*Piper, error) {
+	r, w, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	piper := &Piper{
+		r:       r,
+		w:       w,
+		errChan: make(chan error),
+	}
+
+	go func() {
+		_, err := io.Copy(output, r)
+		piper.errChan <- err
+	}()
+
+	return piper, nil
+}
+
+func (piper *Piper) Input() *os.File {
+	return piper.w
+}
+
+func (piper *Piper) Close() (result error) {
+	// Cancel the Goroutine started in New(): ungracefully Windows and gracefully for other platforms
```

Comment:
Maybe separate the logic into two smaller method like `closeGracefully` and `closeUngracefully`? I think it will be easier to follow the logic

Reply:
See https://github.com/cirruslabs/cirrus-ci-agent/pull/179/commits/004328205e2e0c695d5c08d9af97c5f31e3b1e94 for a more straightforward logic.